### PR TITLE
Turn on add-to-app iOS platform unit tests

### PR DESCRIPTION
**Pull request by jmagman**
## Description

add_to_app_tests Cirrus tests have been skipped for awhile due to flakiness (see https://github.com/flutter/flutter/pull/42444).  Let's see if we can catch more leaks.

- Rewrite the EarlGrey integration tests as XCUITests.
- Move the logic and tests from `add_to_app_tests` to `module_test_ios` since that one was a dummy host app that wasn't testing anything, and we don't need two add-to-app host integration test projects.

## Tests

Updated module_test_ios.dart.

## Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change
- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*

### CODE REVIEW

File: dev/integration_tests/ios_host_app/FlutterUITests/FlutterUITests.m [hunk 1]
```diff
@@ -0,0 +1,117 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+@import XCTest;
+
+@interface FlutterUITests : XCTestCase
```
Comment (jmagman) [inline dev/integration_tests/ios_host_app/FlutterUITests/FlutterUITests.m:7]: This is a rewrite of FlutterViewControllerTests.

File: dev/integration_tests/ios_host_app/Host/AppDelegate.m [hunk 2]
```diff
@@ -3,6 +3,51 @@
 // found in the LICENSE file.
 
 #import "AppDelegate.h"
+#import "MainViewController.h"
```
Comment (jmagman) [inline dev/integration_tests/ios_host_app/Host/AppDelegate.m:6]: This was moved from [dev/integration_tests/ios_add2app/ios_add2app/AppDelegate.m](https://github.com/flutter/flutter/pull/61007/files#diff-9e149ff1c1dcf4b51edd1f16e15eea54) but git didn't follow the rename.

File: dev/integration_tests/ios_host_app/Host/AppDelegate.h [hunk 3]
```diff
@@ -6,4 +6,8 @@
 #import <Flutter/Flutter.h>
 
 @interface AppDelegate : FlutterAppDelegate
+
+@property(nonatomic, strong) FlutterEngine* engine;
```
Comment (jmagman) [inline dev/integration_tests/ios_host_app/Host/AppDelegate.h:10]: This was moved from [dev/integration_tests/ios_add2app/ios_add2app/AppDelegate.h](https://github.com/flutter/flutter/pull/61007/files#diff-857228bf10c8c1e20cb51a974a25baa0) but git didn't follow the rename.

File: dev/integration_tests/ios_host_app/Host/Base.lproj/LaunchScreen.storyboard [hunk 4]
```diff
@@ -1,7 +1,10 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="13122.16" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="01J-lp-oVM">
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jmagman) [inline dev/integration_tests/ios_host_app/Host/Base.lproj/LaunchScreen.storyboard:1]: This was moved from [dev/integration_tests/ios_add2app/ios_add2app/Launch Screen.storyboard](https://github.com/flutter/flutter/pull/61007/files#diff-95f972d78b529be67ee75548a2086d30) but git didn't follow the rename.

File: dev/integration_tests/ios_host_app/Host/MainViewController.m [hunk 5]
```diff
@@ -23,7 +23,7 @@ - (void)viewDidLoad {
   [super viewDidLoad];
 
   [self.view setFrame:self.view.window.bounds];
-  self.title = @"Flutter iOS Demos";
+  self.title = @"Flutter iOS Demos Home";
```
Comment (jmagman) [inline dev/integration_tests/ios_host_app/Host/MainViewController.m:26]: I wanted something identifiable as the home page.

File: dev/integration_tests/ios_host_app/README.md [hunk 6]
```diff
@@ -1,8 +1,34 @@
...
 ```
 $ flutter create -t module hello
 ```
 and placed in a sibling folder to (a clone of) the host app.
-Used by the `module_test_ios.dart` device lab test.
+`flutterapp/lib/marquee.dart` and `flutterapp/lib/main.dart`
+must be copied into the new module `lib` for platform unit tests
+to pass.
+
+This application demonstrates some basic functionality for Add2App,
```
Comment (jmagman) [inline dev/integration_tests/ios_host_app/README.md:14]: Mostly copied from [dev/integration_tests/ios_add2app/README.md](https://github.com/flutter/flutter/pull/61007/files#diff-14972bc6e76cd3b7d00d54d922a9c95f).

File: dev/devicelab/lib/framework/ios.dart [hunk 7]
```diff
@@ -99,3 +101,85 @@ Future<bool> containsBitcode(String pathToBinary) async {
   });
   return !emptyBitcodeMarkerFound;
 }
+
+/// Creates and boots a new simulator, passes the new simulator's identifier to
+/// `testFunction`, then shuts down and deletes simulator.
+Future<void> testWithNewiOSSimulator(
```
Comment (jmagman) [inline dev/devicelab/lib/framework/ios.dart:107]: This was essentially copied from https://github.com/flutter/flutter/blob/023532d9919463404e0ae2a8aa47d07247ad7e65/dev/devicelab/bin/tasks/ios_app_with_extensions_test.dart#L79.  Next PR will convert that test to use this new function.

File: dev/devicelab/bin/tasks/module_test_ios.dart [hunk 8]
```diff
@@ -181,42 +189,6 @@ Future<void> main() async {
...
-
-      await inDirectory(projectDir, () async {
-        await flutter('clean');
-      });
-
-      section('Make iOS host app editable');
-
-      await inDirectory(projectDir, () async {
-        await flutter(
-          'make-host-app-editable',
```
Comment (jmagman) [inline dev/devicelab/bin/tasks/module_test_ios.dart:194]: This was deprecated with https://github.com/flutter/flutter/issues/54408 and is a no-op.

File: dev/integration_tests/ios_add2app/build_and_test.sh [hunk 9]
```diff
@@ -1,26 +0,0 @@
-#!/usr/bin/env bash
```
Comment (jmagman) [inline dev/integration_tests/ios_add2app/build_and_test.sh:1]: Woo one less bash script!

**Review by christopherfujino — APPROVED** [review]
LGTM

diffstat

File: .cirrus.yml [hunk 10]
```diff
@@ -272,8 +272,6 @@ task:
     - name: hostonly_devicelab_tests-3_last-linux
       << : *LINUX_SHARD_TEMPLATE
 
-    # TODO(ianh): name: add_to_app_tests-linux
```
Comment (dnfield) [inline .cirrus.yml:275]: Intentional?

File: .cirrus.yml [hunk 11]
```diff
@@ -508,8 +506,6 @@ task:
         CPU: 4
         MEMORY: 6G
 
-    # TODO(ianh): name: add_to_app_tests-windows
```
Comment (dnfield) [inline .cirrus.yml:511]: ditto

File: dev/integration_tests/ios_host_app/FlutterUITests/FlutterUITests.m (hunk 1 above)
Comment (dnfield) [reply dev/integration_tests/ios_host_app/FlutterUITests/FlutterUITests.m:7]: FWIW, EarlyGrey is about to publish a Flutter plugin.

**Review by dnfield — APPROVED** [review]
LGTM modulo removal of seemingly unrelated TODOs

File: .cirrus.yml (hunk 10 above)
Comment (dnfield) [reply .cirrus.yml:275]: Oh, I get it now, these would correspond to the one you're removing for other platforms. LGTM.

Comment (jmagman) [reply .cirrus.yml:275]: There's no more add_to_app_tests on any platform, so we don't need a windows or linux one.

**Comment by jmagman** [comment]
Added an exclusion to `analysis_options` since the `dev/integration_tests/ios_host_app/flutterapp/lib` dart files have errors on their own, and they don't have a pubspec.yaml.  They are copied into the newly created module during the integration test.
```
  error • The name 'Icon' isn't a class • dev/integration_tests/ios_host_app/flutterapp/lib/main.dart:191:24 • creation_with_non_type
  error • Undefined name 'Icons' • dev/integration_tests/ios_host_app/flutterapp/lib/main.dart:191:29 • undefined_identifier
  error • Target of URI doesn't exist: 'package:flutter/material.dart' • dev/integration_tests/ios_host_app/flutterapp/lib/marquee.dart:5:8 • uri_does_not_exist
  error • Target of URI doesn't exist: 'package:flutter/animation.dart' • dev/integration_tests/ios_host_app/flutterapp/lib/marquee.dart:6:8 • uri_does_not_exist
  error • Target of URI doesn't exist: 'package:flutter/services.dart' • dev/integration_tests/ios_host_app/flutterapp/lib/marquee.dart:7:8 • uri_does_not_exist
  error • Classes can only extend other classes • dev/integration_tests/ios_host_app/flutterapp/lib/marquee.dart:9:28 • extends_non_class
  error • Undefined class 'Key' • dev/integration_tests/ios_host_app/flutterapp/lib/marquee.dart:10:23 • undefined_class
```
etc

File: dev/integration_tests/ios_host_app/analysis_options.yaml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+# These dart files are copied into a newly created project and have
+# errors outside that context.
+analyzer:
+  exclude:
+    - 'flutterapp/lib/**'
```
Comment (jmagman) [inline dev/integration_tests/ios_host_app/analysis_options.yaml:5]: Hm, this isn't working, I'm still getting errors from these files.  You guys have any hints to get it to ignore `ios_host_app/flutterapp/lib/main.dart` and `ios_host_app/flutterapp/lib/marquee.dart`?

Comment (christopherfujino) [reply dev/integration_tests/ios_host_app/analysis_options.yaml]: how about `**/flutterapp/lib/**`?

Comment (jmagman) [reply dev/integration_tests/ios_host_app/analysis_options.yaml]: I think you were looking at an outdated diff.
Also `./dev/integration_tests/ios_add2app_life_cycle/flutterapp/lib` exists and is a real package.

**Comment by jmagman** [comment]
Ugh, didn't even get to the analyzer this time.  Still trying to figure out how to have floating dart test files that aren't really in a package, but will be copied into a newly created package by an integration test.
```
Warning: pubspec in /tmp/flutter sdk/dev/integration_tests/ios_host_app/flutterapp has out of date dependencies. Please run "flutter update-packages --force-upgrade" to update them correctly.
```

File: dev/integration_tests/ios_host_app/analysis_options.yaml (hunk 12 above)
Comment (christopherfujino) [reply dev/integration_tests/ios_host_app/analysis_options.yaml]: > Ugh, didn't even get to the analyzer this time. Still trying to figure out how to have floating dart test files that aren't really in a package, but will be copied into a newly created package by an integration test.
> 
> ```
> Warning: pubspec in /tmp/flutter sdk/dev/integration_tests/ios_host_app/flutterapp has out of date dependencies. Please run "flutter update-packages --force-upgrade" to update them correctly.
> ```

I don't think that test is even using the analyzer, it's just checking the pubspec checksum

Comment (christopherfujino) [reply dev/integration_tests/ios_host_app/analysis_options.yaml]: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/update_packages.dart#L180

**Comment by iprox** [comment]
Hey folks, i stumbled upon this MR and specifically this file

https://github.com/jmagman/flutter/blob/af942cc85348ff24598b6bbc9101663f24c5f999/dev/integration_tests/ios_host_app/FlutterUITests/FlutterUITests.m

I haven't seen any documentation regarding using XCUITests with Flutter add-to-app. Is it something officially supported? Or is there any other place i can look at to understand how we can write integration tests for add-to-app scenario?
Thanks!

**Comment by jmagman** [comment]
@iprox `XCUITests` are totally black-box tests and can't really be framework-aware (unlike [integration_test](https://github.com/flutter/flutter/tree/master/packages/integration_test#ios-device-testing) which requires a non-UI `XCTest` target to validate plugin channel messages).  You should be able to create UI tests like the ones you found in https://github.com/flutter/flutter/blob/f123f64deb1061fcaf15d6a5153ea16e9aeec763/dev/integration_tests/ios_host_app/FlutterUITests/FlutterUITests.m to tap buttons, validate text, etc, since Flutter has accessibility hooked up (if you see any issues, please file them!).

tl;dr there's no official Flutter docs, but it should look the same as any other XCUITest for a non-Flutter app.

**Comment by iprox** [comment]
@jmagman thanks for the reply. I managed to write some simple UI tests, will definitely explore more. As we're using Flutter as add-to-app this is the only way for us to test the flow end-to-end.

**Comment by jmagman** [comment]
@iprox You should also be able to write `XCTest` (non-UI tests).  From the host app's perspective Flutter and the plugins are frameworks like any other.  As long as your test target links on them, it can poke at them the same any framework you embed in your host app.  Most of the complication related to add-to-app is related to compiling your dart code (with build modes, flavors, dart variables, etc).  Once it's built, it's just a bunch of frameworks.
